torch/distributed/tensor/parallel/style.py

```diff
@@ -1,652 +1,357 @@
-                Shard(-1),
-            )
+        super().__init__()
+        self.input_layouts = (input_layouts or Replicate(), )
+        self.output_layouts = (output_layouts or Shard(-1), )
+        # colwise linear runtime sharding (desired sharding):
+        # 1. requires replicate input
+        # 2. shard output on last dim
+        self.desired_input_layouts = (Replicate(), )
+        self.desired_output_layouts = (Shard(-1), )
+        self.use_local_output = use_local_output
 
-        prepare_output_fn = None
-        if _prepare_output is not None:
-            prepare_output_fn = _prepare_output
-        elif output_layouts is not None:
-            prepare_output_fn = functools.partial(
-                _redistribute_per_layout, output_layouts, use_local_output
+    def _prepare_input_fn(self, inputs, device_mesh):
+        # annotate module input placements/sharding with input_layouts
+        dtensor = DTensor.from_local(inputs[0], device_mesh, self.input_layouts)
+        # transform the input layouts to the desired layouts of ColwiseParallel
+        desired_sharding = (Replicate(), )
+        if self.input_layouts != desired_sharding:
+            dtensor = dtensor.redistribute(placements=self.desired_input_layouts)
+        return dtensor
+
+    def _partition_fn(self, name, module, device_mesh):
+        if isinstance(module, nn.Linear):
+            # colwise shard weight/bias to Shard(0), weight be Shard(0)
+            # means Colwise as Linear is input * weight^T + bias, where
+            # weight would become Shard(1)
+            for name, param in module.named_parameters():
+                dist_param = nn.Parameter(
+                    distribute_tensor(param, device_mesh, [Shard(0)])
+                )
+                module.register_parameter(name, dist_param)
+        elif isinstance(module, nn.Embedding):
+            # colwise shard embedding.weight is straight forward as Shard(1)
+            for name, param in module.named_parameters():
+                dist_param = nn.Parameter(
+                    distribute_tensor(param, device_mesh, [Shard(1)])
+                )
+                module.register_parameter(name, dist_param)
+        else:
+            raise NotImplementedError(
+                "ColwiseParallel only supports nn.Linear"
+                f"and nn.Embedding for now, but found {type(module)}!"
             )
 
-        super().__init__(
-            input_layouts=input_layouts,
-            output_layouts=output_layouts,
-            use_local_output=use_local_output,
-            _prepare_input=prepare_input_fn,
-            _prepare_output=prepare_output_fn,
-        )
-
-    @staticmethod
-    def _prepare_input_fn(input_layouts, output_layouts, inputs, device_mesh=None):
-        return (
-            _redistribute_per_both_layouts(
-                inputs[0], input_layouts, output_layouts, device_mesh
-            ),
+    def _prepare_output_fn(self, outputs, device_mesh):
```

Comment:
Super weird that the Python highlight is not rendered correctly here, but when I clicked on view file, everything seems ok.